File: a1/Prolog/rush_hour/state.pl
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% rush_hour/state.pl
%
% Rush Hour puzzle solver
% (C) 2019 Norbert Zeh (nzeh@cs.dal.ca)
%
% Predicates to manipulate board states
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% Translate a string description of the puzzle into a list of 4 64-bit integers
% encoding which grid cells are occupied, by horizontal pieces, by vertical
% pieces, and by the rightmost cell of a horizontal piece or the bottommost
% cell of a vertical piece.
puzzle_state(Puzzle, State) :-
  puzzle_grid(Puzzle, Grid),
  grid_state(Grid, State).


% Pad a 36-character description of a board into an 8x8 grid of characters that
% includes empty cells where the walls are.
puzzle_grid(Puzzle, Grid) :-
  string_chars(Puzzle, Chars),
  string_chars(oooooooo, EmptyRow),
  Rows = [_ ,_ ,_ ,_ ,_ ,_],
  chars_rows(Chars, Rows),
  append([[EmptyRow], Rows, [EmptyRow]], Grid).


% Unpack the rows of the a grid from the 36-character representation
chars_rows([], []).

chars_rows([A, B, C, D, E, F | Rest], [[o, A, B, C, D, E, F, o] | Rows]) :-
  chars_rows(Rest, Rows).


% Transpose a grid
grid_transposed([], [[], [], [], [], [], [], [], []]).

grid_transposed([Row|Rows], Transposed) :-
  grid_transposed(Rows, Cols),
  maplist(list_head_tail, Transposed, Row, Cols).


% True if the second and third arguments are the head and tail of the first
% argument.
list_head_tail([X|Xs], X, Xs).


% Translate an 8x8 grid into a state object
grid_state(Grid, [Occupied, Horiz, Vert, Ends]) :-
  grid_transposed(Grid, Transposed),
  grid_horiz_ends(Grid, Horiz, HorizEnds),
  grid_horiz_ends(Transposed, VertTransposed, VertEndsTransposed),
  bits_transposed(VertTransposed, Vert),
  bits_transposed(VertEndsTransposed, VertEnds),
  Occupied is Horiz \/ Vert,
  Ends is HorizEnds \/ VertEnds.


% Rearrange the bits in a 64-bit string to model transposition of an 8x8 bit
% matrix.
bits_transposed(Bits, Transposed) :-
  T0 is (Bits xor (Bits << 7)) /\ 0x5500550055005500,
  B1 is Bits xor T0 xor (T0 >> 7),
  T1 is (B1 xor (B1 << 14)) /\ 0x3333000033330000,
  B2 is B1 xor T1 xor (T1 >> 14),
  T2 is (B2 xor (B2 << 28)) /\ 0x0f0f0f0f00000000,
  Transposed is B2 xor T2 xor (T2 >> 28).


% Locate all horizontal pieces in a grid and construct two 64-bit numbers that
% represent the cells occupied by these pieces and their rightmost cells.
grid_horiz_ends(Grid, Horiz, Ends) :-
  append(Grid, Cells),
  accumulate_horiz_ends(o, Cells, 0xff000000000000ff, 0, 1, Horiz, Ends).


% Inspect the cells in a list one by one to construct the corresponding bit
% vectors.
accumulate_horiz_ends(_, [], Bits, Ends, _, Bits, Ends) :- !.

accumulate_horiz_ends(_, [o|Xs], BitsIn, EndsIn, PosIn, BitsOut, EndsOut) :-
  PosN is PosIn << 1, !,
  accumulate_horiz_ends(o, Xs, BitsIn, EndsIn, PosN, BitsOut, EndsOut).

accumulate_horiz_ends(_, [X,X|Xs], BitsIn, EndsIn, PosIn, BitsOut, EndsOut) :-
  PosN is PosIn << 1,
  BitsN is BitsIn \/ PosIn, !,
  accumulate_horiz_ends(X, [X|Xs], BitsN, EndsIn, PosN, BitsOut, EndsOut).

accumulate_horiz_ends(X, [X|Xs], BitsIn, EndsIn, PosIn, BitsOut, EndsOut) :-
  PosN is PosIn << 1,
  BitsN is BitsIn \/ PosIn,
  EndsN is EndsIn \/ PosIn, !,
  accumulate_horiz_ends(X, Xs, BitsN, EndsN, PosN, BitsOut, EndsOut).

accumulate_horiz_ends(_, [X|Xs], BitsIn, EndsIn, PosIn, BitsOut, EndsOut) :-
  PosN is PosIn << 1, !,
  accumulate_horiz_ends(X, Xs, BitsIn, EndsIn, PosN, BitsOut, EndsOut).


% Check whether the given state is solved
state_is_solved(State) :-
  state_is_horizontal(State, 30).


% Check whether the Posth grid cell is occupied
state_is_ocupied([Occupied, _, _, _], Pos) :-
  state_bit_is_set(Occupied, Pos).


% Check whether the Posth grid cell is occupied by a horizontal piece
state_is_horizontal([_, Horiz, _, _], Pos) :-
  state_bit_is_set(Horiz, Pos).


% Check whether the Posth grid cell is occupied by a vertical piece
state_is_vertical([_, _, Vert, _], Pos) :-
  state_bit_is_set(Vert, Pos).


% Check whether the Posth grid cell is the rightmost cell of a horizontal piece
% or the bottommost cell of a vertical piece
state_is_end([_, _, _, Ends], Pos) :-
  state_bit_is_set(Ends, Pos).


% Check that the Posth bit in Bits is set
state_bit_is_set(Bits, Pos) :-
  Bit is Bits /\ (1 << Pos),
  Bit \= 0.


% Given a state, a position, and an offset, the predicate is true if and only
% if position Pos is the rightmost cell of a horizontal piece, this piece can
% be moved horizontally by Offset positions, and NewState is the resulting
% state.  Should be used as a function to compute NewState from State, Pos, and
% Offset by passing a variable for NewState.
horizontal_move(State, Pos, Offset, NewState) :-
  state_is_end(State, Pos), state_is_horizontal(State, Pos),
  State = [Occupied, _, Vert, Ends],
  Stop is Occupied
       /\ (Vert xor 0xffffffffffffffff)
       /\ (Ends xor 0xffffffffffffffff),
  move(State, Pos, Offset, Stop, 1, 0xffffffffffffffff, NewState).


% Given a state, a position, and an offset, the predicate is true if and only
% if position Pos is the bottommost cell of a vertical piece, this piece can be
% moved vertical by Offset positions, and NewState is the resulting state.
% Should be used as a function to compute NewState from State, Pos, and Offset
% by passing a variable for NewState.
vertical_move(State, Pos, Offset, NewState) :-
  state_is_end(State, Pos), state_is_vertical(State, Pos),
  State = [Occupied, Horiz, _, Ends],
  Stop is Occupied
       /\ (Horiz xor 0xffffffffffffffff)
       /\ (Ends xor 0xffffffffffffffff),
  move(State, Pos, Offset, Stop, 8, 0x0101010101010101, NewState).


% The actual worker predicate used to implement horizontal_move and
% vertical_move.
move([Occupied, Horiz, Vert, Ends],
     Pos, Offset, Stop, Skip, Mask,
     [NewOccupied, NewHoriz, NewVert, NewEnds]) :-

  % Locate the piece and calculate its new position
  find_piece(Stop, Pos, Skip, Piece),
  NewPiece is Piece << (Offset * Skip),
  NewPos is Pos + (Offset * Skip),

  % Any position outside the range 9-54 is outside the board and is not
  % permissible, so the predicate fails.
  NewPos >= 9, NewPos =< 54,

  % Figure out the swath of positions over which the piece moves from its old
  % to its new position
  Left is min(Piece, NewPiece),
  Right is max(Piece, NewPiece),
  LeftPos is min(Pos, NewPos),
  RightPos is max(Pos, NewPos),
  LeftMask is Left \/ (Mask << LeftPos),
  RightMask is Right \/ (Mask >> (64 - RightPos - Skip)),
  Swath is LeftMask /\ RightMask,

  % The swath of positions over which the piece moves should not be occupied by
  % any other piece
  Conflicts is Swath /\ (Occupied xor Piece),
  Conflicts = 0,

  % Now that we know the move is legal, construct the new state.
  PieceDelta is Piece xor NewPiece,
  EndsDelta is (1 << Pos) \/ (1 << (Pos + Offset * Skip)),
  NewOccupied is Occupied xor PieceDelta,
  NewEnds is Ends xor EndsDelta,
  (Skip == 1 -> (HorizDelta, VertDelta) = (PieceDelta, 0); (HorizDelta, VertDelta) = (0, PieceDelta)),
  NewHoriz is Horiz xor HorizDelta,
  NewVert is Vert xor VertDelta.


% Find the piece whose bottommost or rightmost position is Pos.
find_piece(Stop, Pos, Skip, Piece) :-
  Init is 1 << Pos,
  Next is Init >> Skip,
  find_piece0(Stop, Skip, Init, Next, Piece).


% Helper that implements the recursion of looking for the piece
find_piece0(Stop, Skip, Init, Next, Piece) :-
  More is Stop /\ Next,
  ( More = 0 -> Piece = Init;
    Init0 is Init \/ Next,
    Next0 is Next >> Skip,
    find_piece0(Stop, Skip, Init0, Next0, Piece)
  ).


% Construct a move object from a position and an offset
pos_offset_move(Pos, Offset, Move) :-
  Move is (Pos << 8) \/ (Offset + 4).


% Format the move into a string
move_string(Move, String) :-
  with_output_to(string(String), format_move(Move)).

format_move(Move) :-
  Pos is Move >> 8,
  Offset is (Move /\ 15) - 4,
  Row is Pos >> 3,
  Col is Pos /\ 7,
  Abs is abs(Offset),
  (Offset == Abs -> Sign = "+"; Sign = "-"),
  format("(~d,~d)~s~d~n", [Row, Col, Sign, Abs]).

File: a1/Python/rush_hour/state.py
###############################################################################
#
# rush_hour/state.py
#
# Rush Hour puzzle solver
# (C) 2019 Norbert Zeh (nzeh@cs.dal.ca)
#
###############################################################################


"""This module provides the primitives for manipulating and querying the
current game state.

The game state is represented as a tuple of 4 64-bit words
(occupied, horiz, vert, ends).  Each bit corresponds to a cell of an 8x8 board
numbered in row-major order.  Referring to the ith bit of a bit vector vec as
vec[i], these four words have the following meaning:

- occupied[i] = 1 if cell i is occupied by a piece
- horiz[i] = 1    if cell i is occupied by a horizontal piece
- vert[i] = 1     if cell i is occupied by a vertical piece
- ends[i] = 1     if cell i is the rightmost or bottommost cell of a piece"""


from io import StringIO
from itertools import chain


def from_string_rep(string_rep):
    """Construct a state from its string representation.  The input is a
    36-character string listing the 6x6 cells of the board in row-major order.
    Empty cells are marked with "o".  Occupied cells are marked with letters
    representing pieces.  Cells occupied by the same piece carry the same
    letter."""
    grid = _grid_from_string_rep(string_rep)
    horiz, horiz_ends = _find_horiz_pieces(grid)
    vert, vert_ends = _find_vert_pieces(grid)
    occupied = vert | horiz
    ends = horiz_ends | vert_ends
    return (occupied, horiz, vert, ends)


def _grid_from_string_rep(string_rep):
    """Pad a 36-character string representation of a state into a 64-character
    representation that includes empty cells to represent the borders around
    the board."""
    grid = StringIO()
    print("oooooooo", file=grid, end="")
    for i in range(6):
        print("o{}o".format(string_rep[6*i:6*(i+1)]), file=grid, end="")
    print("oooooooo", file=grid, end="")
    return grid.getvalue()


def _find_vert_pieces(grid):
    """Find all vertical pieces on the board and return two words representing
    the cells occupied by these pieces and the cells that are the bottommost
    cells of these pieces."""
    transposed_grid = _transpose_grid(grid)
    transposed_bits, transposed_ends = \
        _find_horiz_pieces(transposed_grid)
    return _transpose_bits(transposed_bits), \
        _transpose_bits(transposed_ends)


def _find_horiz_pieces(grid):
    """Find all horizontal pieces on the board and return two words
    representing the cells occupied by these pieces and the cells that are the
    rightmost cells of these pieces."""
    bits = 0xff000000000000ff
    ends = 0
    for i in range(64):
        cellbit, cellend = _horiz_bits(grid, i)
        bits |= cellbit
        ends |= cellend
    return bits, ends


def _horiz_bits(grid, pos):
    """Return a pair of words representing position pos on the board.  The
    first word is 1 << pos if the position is occupied and 0 otherwise.  The
    second word is 1 << pos if the position is the rightmost position of a
    piece and 0 otherwise."""
    if grid[pos] == "o":
        return 0, 0
    if grid[pos] == grid[pos+1]:
        return 1 << pos, 0
    if grid[pos] == grid[pos-1]:
        return 1 << pos, 1 << pos
    else:
        return 0, 0


def _transpose_grid(grid):
    """Rearranges the character in a 64-character string represending an 8x8
    Rush Hour board so as to transpose the board."""
    transposed_grid = StringIO()
    for i in range(64):
        j = ((i & 7) << 3) | ((i & 56) >> 3)
        print(grid[j], file=transposed_grid, end="")
    return transposed_grid.getvalue()


def _transpose_bits(bits):
    """Rearranges the bits in a 64-bit bit string representing an 8x8 board so
    as to transpose the board."""
    # 2x2 transpose of individual bits
    trans = (bits ^ (bits << 7)) & 0x5500550055005500
    bits = bits ^ trans ^ (trans >> 7)
    # 2x2 transpose of 2x2 blocks
    trans = (bits ^ (bits << 14)) & 0x3333000033330000
    bits = bits ^ trans ^ (trans >> 14)
    # 2x2 transpose of 4x4 blocks
    trans = (bits ^ (bits << 28)) & 0x0f0f0f0f00000000
    bits = bits ^ trans ^ (trans >> 28)
    return bits


def pretty_print(state):
    """Produce a vector of 14 14-character strings that, if printed in
    consecutive rows display the Rush Hour board.  (The 14 rows and 14 columns
    result from using 2x2 characters for every board cell and one character for
    the border.  Since there are 6x6 cells on the board, this gives 14 rows and
    14 columns.)"""
    rows = [[_format_cell(state, row + col) for col in range(1, 7)]
            for row in range(8, 63, 8)]
    lines = ["████████████"] + \
            ["".join(chain.from_iterable(line))
             for line in chain.from_iterable((zip(*row) for row in rows))] + \
            ["████████████"]
    left_wall = "██████████████"
    right_wall = "█████  ███████"
    return ["".join(line) for line in zip(left_wall, lines, right_wall)]


def _format_cell(state, pos):
    if not is_occupied(state, pos):
        return ["  ", "  "]
    if is_horizontal(state, pos):
        if not is_horizontal(state, pos - 1) or is_end(state, pos - 1):
            left = "▗▝"
        else:
            left = "▄▀"
        if not is_horizontal(state, pos + 1) or is_end(state, pos):
            right = "▖▘"
        else:
            right = "▄▀"
        return ["".join(line) for line in zip(left, right)]
    if not is_vertical(state, pos - 8) or is_end(state, pos - 8):
        top = "▗▖"
    else:
        top = "▐▌"
    if not is_vertical(state, pos + 8) or is_end(state, pos):
        bottom = "▝▘"
    else:
        bottom = "▐▌"
    return [top, bottom]


def is_occupied(state, pos):
    """Check whether the posth cell is occupied."""
    return state[0] & (1 << pos)


def is_horizontal(state, pos):
    """Check whether the posth cell is occupied by a horizontal piece."""
    return state[1] & (1 << pos)


def is_vertical(state, pos):
    """Check whether the posth cell is occupied by a vertical piece."""
    return state[2] & (1 << pos)


def is_end(state, pos):
    """Check whether the posth cell is the rightmost or bottommost cell of a
    piece."""
    return state[3] & (1 << pos)


def is_solved(state):
    """Check whether the board is solved.  This is the case when the horizontal
    on the 3rd row is moved all the way to the right, that is cell 30 is
    occupied by a horizontal piece."""
    return is_horizontal(state, 30)


def make_move(pos, offset):
    """Construct a move object from a given position-offset pair"""
    return (pos << 8) | (offset + 4)


def vertical_move(state, pos, offset):
    """Move the vertical piece occupying position pos by offset positions.  A
    negative offset means move up.  A positive offset means move down.  The
    return value is the new state and a representation of the move as a 16-bit
    word.  If the move is invalid because it moves the piece off the board or
    across another piece, two None values are returned."""
    stop = state[0] & \
        (state[1] ^ 0xffffffffffffffff) & \
        (state[3] ^ 0xffffffffffffffff)
    return _move(state, pos, offset, stop, 8, 0x0101010101010101)


def horizontal_move(state, pos, offset):
    """Move the horizontal piece occupying position pos by offset positions.  A
    negative offset means move left.  A positive offset means move right.  The
    return value is the new state and a representation of the move as a 16-bit
    word.  If the move is invalid because it moves the piece off the board or
    across another piece, two None values are returned."""
    stop = state[0] & \
        (state[2] ^ 0xffffffffffffffff) & \
        (state[3] ^ 0xffffffffffffffff)
    return _move(state, pos, offset, stop, 1, 0xffffffffffffffff)


def _move(state, pos, offset, stop, skip, mask):
    """This is the worker that implements both horizontal and vertical
    moves."""

    # Construct the bit vector representing the positions occupied by the piece
    # to be moved
    piece = 1 << pos
    left = piece >> skip
    while stop & left:
        piece |= left
        left >>= skip

    # Construct the new piece resulting from the move and
    # - the leftmost and rightmost of piece and new_piece
    # - the end positions of the leftmost and rightmost pieces
    if offset < 0:
        new_piece = piece >> (-offset * skip)
        left, right = new_piece, piece
        left_pos, right_pos = pos + offset * skip, pos
    else:
        new_piece = piece << (offset * skip)
        left, right = piece, new_piece
        left_pos, right_pos = pos, pos + offset * skip

    # Abort if the new piece is out of bounds
    if left_pos < 9 + skip or right_pos > 54:
        return None

    # Construct the swath of cells over which the piece moves from its old to
    # its new position.
    left = left | (mask << left_pos)
    right = right | (mask >> (64 - right_pos - skip))
    swath = (left & right) ^ piece

    # Abort if any of them is occupied.
    if state[0] & swath:
        return None

    # Construct the change in occupied cells and the change in end cells
    # resulting from the move
    piece_delta = piece ^ new_piece
    end_delta = (1 << pos) | (1 << (pos + offset * skip))

    # Update occupied, horiz, and end if this was a horizontal move.
    # Otherwise, update, occupied, vert, and end.
    if skip == 1:
        return (
            state[0] ^ piece_delta,
            state[1] ^ piece_delta,
            state[2],
            state[3] ^ end_delta
        )
    return (
        state[0] ^ piece_delta,
        state[1],
        state[2] ^ piece_delta,
        state[3] ^ end_delta
    )


def apply_move(state, move):
    """Try to apply the given move to state.  Return the new state if this
    succeeds.  Otherwise, return None."""
    pos = move >> 8
    offset = (move & 0xff) - 4
    if not is_occupied(state, pos):
        return None
    if is_horizontal(state, pos):
        return horizontal_move(state, pos, offset)
    return vertical_move(state, pos, offset)
